sip/sip-002-smart-contract-language.md

```diff
@@ -0,0 +1,309 @@
+
+1. The language _is not_ intended to be compiled. The LISP language
+   described in this document is the specification for correctness.
+2. The language _is not_ Turing complete. This allows us to guarantee
+   that static analysis of programs to determine properties like
+   runtime cost and data usage can complete successfully.
+
+## Specifying Transactions
+
+A smart contract definition is specified in a LISP language with the
+following limitations:
+
+1. Recursion is illegal and there is no `lambda` function.
+2. Looping may only be performed via `map`, `filter`, or `fold`
+3. The only atomic types are booleans, integers, fixed length
+   buffers, and principals
+4. There is additional support for lists of the atomic types,
+   however the only variable length lists in the language appear as
+   transaction inputs (i.e., there is no support for list operations
+   like append or join).
+5. Variables may only be created via `let` binding and there
+   is no support for mutating functions like `set`.
+6. Defining of constants and functions are allowed for simplifying
+   code. However, these are purely syntactic. If a definition cannot be
+   inlined, the contract will be rejected as illegal.
+7. Transactions are specified via `define` statement with function
+   names beginning with `tx-`. Arguments to the function must specify
+   their types.
+
+If a transaction returns `true`, then it is considered valid, and any
+changes made to the blockchain state will be materialized. If a
+transaction returns `false`, the transaction will be considered
+invalid, and the transaction will have _no effect_ on the smart
+contract's state, except for a transaction fee debit (in the case of
+on-chain transactions).
+
+## Verifying Signing Principals
+
+The language provides a primitive for checking whether or not the
+smart contract transaction was signed by a particular
+_principal_. Principals are a representation of a signing entity
+(roughly equivalent to a Stacks address). The signature itself is
+not checked by the smart contract, but by the VM. To check whether a
+given principal has signed the transaction, a transaction may call
+
+```scheme
+(signed-by? principal)
+```
+
+This returns `true` or `false`. Importantly, to support inter-contract
+calls, this function returns `true` if the _outermost_ transaction was
+signed by the given principal. A key benefit of the static
+analyzability of this smart contracting language is that _all_
+transactions that can possibly be called from the outermost
+transaction can be known _a priori_ so that a user can be warned about
+all side effects before signing a transaction.
+
+## Stacks Transaction Primitives
+
+To interact with Stacks balances, smart contracts may call the
+`(stacks-transfer!)` function. This function will attempt to transfer
+from a given principal to another principal. This function itself
+_requires_ that the operation have been signed by the transfering
```

Comment:
Why even have a parameter for the principle that is transferring if there is only one possible valid value for this parameter?

I suppose it makes static analysis a lot easier.

Reply:
Fair point -- I think if the transaction "sender" is always an implied parameter (or specifically, there's an always defined variable `tx-sender`), then transaction definitions don't need the sender as a parameter.

Reply:
Updated in latest draft.